File: Enhanced Artifacts/ENHANCED CS 320/Appointment Service/AppointmentService.java
/* 
 * Michael Thomas
 * CS 499
 * 5/25/2025
 * AppointmentService.java
 * Application for adding, deleting, and updating appointments
*/

package mod5.snhu;
import java.util.*;

/**
 * Service class for managing appointments.
 */
public class AppointmentService {
    private final List<Appointment> appointments = new ArrayList<>(); // Stores the list of appointments

    /**
     * Adds a new appointment if the ID is unique.
     * @param appointment Appointment to add
     * @return true if added, false if duplicate ID
     */
    public boolean addAppointment(Appointment appointment) {
        if (appointment == null) {
            throw new IllegalArgumentException("Appointment cannot be null");
        }

        // Check for uniqueness of ID
        for (Appointment existing : appointments) {
            if (existing.getId().equalsIgnoreCase(appointment.getId())) {
                return false; // Duplicate ID found
            }
        }

        appointments.add(appointment); // Add appointment
        return true;
    }

    /**
     * Deletes an appointment by ID.
     * @param id ID of the appointment to delete
     * @return true if deleted, false if not found
     */
    public boolean deleteAppointment(String id) {
        if (id == null || id.isBlank()) {
            throw new IllegalArgumentException("Invalid ID: cannot be null or blank");
        }

        Iterator<Appointment> iterator = appointments.iterator();
        while (iterator.hasNext()) {
            Appointment appointment = iterator.next();
            if (appointment.getId().equalsIgnoreCase(id)) {
                iterator.remove();
                return true; // Successfully removed
            }
        }

        return false; // ID not found
    }
}

File: Enhanced Artifacts/ENHANCED CS 320/Appointment Service/appointment_service.py
"""
Michael Thomas
CS 499
5/25/2025
appointment_service.py
Appointment application for adding, deleting, and updating appointments
"""

from typing import List, Optional
from appointment import Appointment

class AppointmentService:
    def __init__(self):
        self.appointments: List[Appointment] = []

    def add_appointment(self, appointment: Appointment) -> bool:
        # Ensure appointment ID is unique
        if any(existing.get_id() == appointment.get_id() for existing in self.appointments):
            return False
        self.appointments.append(appointment)
        return True

    def delete_appointment(self, appointment_id: str) -> bool:
        for appointment in self.appointments:
            if appointment.get_id() == appointment_id:
                self.appointments.remove(appointment)
                return True
        return False

    def update_appointment(self, appointment_id: str, new_date=None, new_description=None) -> bool:
        # Update the appointment with new date and/or description if it exists
        for appointment in self.appointments:
            if appointment.get_id() == appointment_id:
                if new_date:
                    appointment.set_date(new_date)
                if new_description:
                    appointment.set_description(new_description)
                return True
        return False

    def get_appointment_by_id(self, appointment_id: str) -> Optional[Appointment]:
        # Retrieve an appointment by ID if it exists
        for appointment in self.appointments:
            if appointment.get_id() == appointment_id:
                return appointment
        return None
